genkey test --levels=3 --lms=15 --lmots=1 --alg=sha256 --trunc=32

This is failing because it supports 0x200000000000 signatures and genkey tries to write this number as a 32 bit value in the private key file.

Questions:
- I did not find a specific file format in NIST documents nor in rfc8554, do you confirm that the file format is 'custom' rather than 'standardized' ?
- Fixing that 'cleanly' would mean changing the file format for private keys. Do you see an issue with that ?

stack trace using [my fork](https://github.com/sebastien-riou/pyhsslms/commit/8c587ca43c495a94fd9ddebf5451fd1feefa98a5)
```
$ python -m pyhsslms.hsslms genkey test --levels=3 --lms=15 --lmots=1 --alg=sha256 --trunc=32
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\_shared\repos\pyhsslms\pyhsslms\hsslms.py", line 264, in <module>
    main()
  File "C:\_shared\repos\pyhsslms\pyhsslms\hsslms.py", line 187, in main
    pyhsslms.HssLmsPrivateKey.genkey(keyname, levels=levels,
  File "C:\_shared\repos\pyhsslms\pyhsslms\pyhsslms.py", line 1182, in genkey
    prv_file.write(hss_prv.serialize())
                   ^^^^^^^^^^^^^^^^^^^
  File "C:\_shared\repos\pyhsslms\pyhsslms\pyhsslms.py", line 936, in serialize
    assert self._signatures_remaining.bit_length() <= 32, "_signatures_remaining=0x%x, %d bits"%(self._signatures_remaining,self._signatures_remaining.bit_length())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: _signatures_remaining=0x200000000000, 46 bits
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -53,7 +53,23 @@ Revision 1.1.1, released 9-Jan-2023
 Revision 2.0.0, released ???-Oct-2023
 ------------------------------------
 
+- Introduced new HssPrivateKey serialization scheme that allows significantly
+  larger keys to be generated. Previously, only keys that allowed for a total of
+  32-bit length signature uses were possible. This change does not break
+  reverse-compatibility for keys generated using the old serialization scheme.
+  (Contribution by joshuayuen99).
+
+- Correctly generate the LMOTS public key from a given LMOTS private
+  key (Contribution by joshuayuen99).
+
+- Correctly sign messages with LMOTS private keys
+  (Contribution by joshuayuen99).
+
 - Correctly calculate the the size of the LMOTS public key during
   deserialization (Contribution by joshuayuen99).
+
+- Refactor "S" field within LmotsSignature, LmotsPrivateKey, and
+  LmotsPublicKey to "I" and "q" components as per the final RFC
+  (Contribution by joshuayuen99).
 
 - Add tests for the coef function.

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -948,9 +948,10 @@ class HssPrivateKey(object):
         return 2**(self.levels*h)
 
     def serialize(self):
-        assert self._signatures_remaining.bit_length() <= 32, "_signatures_remaining=0x%x, %d bits"%(self._signatures_remaining,self._signatures_remaining.bit_length())
+        # assert self._signatures_remaining.bit_length() <= 32, "_signatures_remaining=0x%x, %d bits"%(self._signatures_remaining,self._signatures_remaining.bit_length())
         
-        rv = u32(self.levels) + u32(0)
+        # make sure to maintain 4-byte padding of all 1's to indicate we should use deserializeV2
+        rv = u32(self.levels) + fromHex('1'*8)
         serialized_prv0 = self.prv[0].serialize()
         rv += u32(len(serialized_prv0)) + serialized_prv0
 
@@ -968,7 +969,7 @@ class HssPrivateKey(object):
             raise ValueError(err_bad_length, str(len(buffer)))
         levels = int32(buffer[0:4])
         rs = int32(buffer[4:8])
-        if rs == 0:
+        if rs == int32(fromHex('1'*8)):
             return cls.deserializeV2(buffer)
         prv = LmsPrivateKey.deserialize(buffer[8:])
         return cls(levels, lms_type=prv.lms_type, lmots_type=prv.lmots_type, \
@@ -978,7 +979,7 @@ class HssPrivateKey(object):
     def deserializeV2(cls, buffer):
         """
         levels - 4 bytes
-        padding of all zeros - 4 bytes
+        padding of all ones - 4 bytes
         [root key_length in bytes + root key]
         [signature_length in bytes + signature + key_length in bytes + key]... for each additional level's private key
         """

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -646,7 +646,7 @@ class LmsPrivateKey(object):
             raise ValueError(err_algorithm_mismatch, alg + ' and ' + alg2)
         SEED = buffer[8:8+n]
         I = buffer[8+n:8+n+LenI]
-        q = int32(buffer[8+n+LenI:8+n+LenI+4])
+        q = int32(buffer[8+n+LenI:8+n+LenI+LenQ])
         return cls(lms_type, lmots_type, SEED, I, q)
 
     def path(self, node_num):

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -145,7 +145,7 @@ lms_params = {
 #
 MaxHssLevels = 8
 LenI = 16
-LenS = LenI + 4
+LenQ = 4
 
 
 # Diversification constants
@@ -290,18 +290,20 @@ class LmotsSignature():
     def serialize(self):
         return self.type + self.C + serialize_list(self.y)
 
-    def buildPublic(self, S, message):
+    def buildPublic(self, I, q, message):
         alg, n, p, w, ls = lmots_params[self.type]
-        if len(S) != LenS:
-            raise ValueError(err_bad_length, str(len(S)))
-        hash1 = H(alg, S + D_MESG + self.C + message, n)
+        if len(I) != LenI:
+            raise ValueError(err_bad_length, str(len(I)))
+        if len(q) != LenQ:
+            raise ValueError(err_bad_length, str(len(q)))
+        hash1 = H(alg, I + q + D_MESG + self.C + message, n)
         V = hash1 + checksum(hash1, w, ls)
         hash = H_start(alg)
-        H_update(hash, S + D_PBLC)
+        H_update(hash, I + q + D_PBLC)
         for i, y in enumerate(self.y):
             tmp = y
             for j in range(coef(V, i, w), (2**w)-1):
-                tmp = H(alg, S + u16(i) + u8(j) + tmp, n)
+                tmp = H(alg, I + q + u16(i) + u8(j) + tmp, n)
             H_update(hash, tmp)
         return H_finish(hash, n)
 
@@ -351,17 +353,21 @@ class LmotsPrivateKey:
     """
     # Generate an LMOTS Private Key
     #
-    def __init__(self, S=None, SEED=None, lmots_type=lmots_sha256_n32_w8):
+    def __init__(self, I=None, q=None, SEED=None, lmots_type=lmots_sha256_n32_w8):
         if lmots_type not in lmots_params:
             raise ValueError(err_unknown_typecode, toHex(lmots_type))
         self.type = lmots_type
         alg, n, p, w, ls = lmots_params[lmots_type]
-        if S is None:
-            self.S = randBytes(LenS)
+        if I is None or q is None:
+            self.I = randBytes(LenI)
+            self.q = randBytes(LenQ)
         else:
-            if len(S) != LenS:
-                raise ValueError(err_bad_length, str(len(S)))
-            self.S = S
+            if len(I) != LenI:
+                raise ValueError(err_bad_length, str(len(I)))
+            if len(q) != LenQ:
+                raise ValueError(err_bad_length, str(len(q)))
+            self.I = I
+            self.q = q
         if SEED is None:
             self.SEED = randBytes(n)
         else:
@@ -379,26 +385,26 @@ class LmotsPrivateKey:
     def publicKey(self): 
         alg, n, p, w, ls = lmots_params[self.type]
         hash = H_start(alg)
-        H_update(hash, self.S + D_PBLC)
+        H_update(hash, self.I + self.q + D_PBLC)
         for i in range(0, p):
-            tmp = H(alg, self.S + u16(i+1) + D_PRG + self.SEED, n)
+            tmp = H(alg, self.I + self.q + u16(i+1) + D_PRG + self.SEED, n)
             for j in range(0, (2**w)-1):
-                tmp = H(alg, self.S + u16(i) + u8(j) + tmp, n)
+                tmp = H(alg, self.I + self.q + u16(i) + u8(j) + tmp, n)
             H_update(hash, tmp)
-        return LmotsPublicKey(self.S, H_finish(hash, n), self.type)
+        return LmotsPublicKey(self.I, self.q, H_finish(hash, n), self.type)
 
     def sign(self, message):
         if self._signatures_remaining != 1:
             raise ValueError(err_private_key_exhausted)
         alg, n, p, w, ls = lmots_params[self.type]
         C = randBytes(n)
-        hash1 = H(alg, self.S + D_MESG + C + message, n)
+        hash1 = H(alg, self.I + self.q + D_MESG + C + message, n)
         V = hash1 + checksum(hash1, w, ls)
         y = []
         for i in range(0, p):
-            tmp = H(alg, self.S + u16(i+1) + D_PRG + self.SEED, n)
+            tmp = H(alg, self.I + self.q + u16(i+1) + D_PRG + self.SEED, n)
             for j in range(0, coef(V, i, w)):
-                tmp = H(alg, self.S + u16(i) + u8(j) + tmp, n)
+                tmp = H(alg, self.I + self.q + u16(i) + u8(j) + tmp, n)
             y.append(tmp)
         self._signatures_remaining = 0
         return LmotsSignature(C, y, self.type).serialize()
@@ -406,7 +412,8 @@ class LmotsPrivateKey:
     def prettyPrint(self):
         rv = "LMOTS private key\n"
         rv += ("   LMOTS type: %s\n" % toHex(self.type))
-        rv += ("   S         : %s\n" % toHex(self.S))
+        rv += ("   I         : %s\n" % toHex(self.I))
+        rv += ("   q         : %s\n" % toHex(self.q))
         rv += ("   SEED      : %s\n" % toHex(self.SEED))
         return rv
 
@@ -415,8 +422,9 @@ class LmotsPublicKey:
     """
     Leighton-Micali One Time Signature Public Key
     """
-    def __init__(self, S, K, lmots_type):
-        self.S = S
+    def __init__(self, I, q, K, lmots_type):
+        self.I = I
+        self.q = q
         self.K = K
         self.type = lmots_type
 
@@ -425,18 +433,18 @@ class LmotsPublicKey:
         if (signature.type != self.type):
             raise ValueError(err_unknown_typecode)
         alg, n, p, w, ls = lmots_params[self.type]
-        hash1 = H(alg, self.S + D_MESG + signature.C + message, n)
+        hash1 = H(alg, self.I + self.q + D_MESG + signature.C + message, n)
         V = hash1 + checksum(hash1, w, ls)
         hash = H_start(alg)
-        H_update(hash, self.S + D_PBLC)
+        H_update(hash, self.I + self.q + D_PBLC)
         for i, tmp in enumerate(signature.y):
             for j in range(coef(V, i, w), (2**w)-1):
-                tmp = H(alg, self.S + u16(i) + u8(j) + tmp, n)
+                tmp = H(alg, self.I + self.q + u16(i) + u8(j) + tmp, n)
             H_update(hash, tmp)
         return self.K == H_finish(hash, n)
 
     def serialize(self):
-        return self.type + self.S + self.K 
+        return self.type + self.I + self.q + self.K 
 
     @classmethod
     def deserialize(cls, buffer):
@@ -446,12 +454,13 @@ class LmotsPublicKey:
         if lmots_type not in lmots_params:
            raise ValueError(err_unknown_typecode, toHex(lmots_type))
         alg, n, p, w, ls = lmots_params[lmots_type]
-        buf_size = 4 + LenS + n
+        buf_size = 4 + LenI + LenQ + n
         if len(buffer) != buf_size:
             raise ValueError(err_bad_length, f"{str(len(buffer))} != {buf_size}")
-        S = buffer[4:4+LenS]
-        K = buffer[4+LenS:(4+LenS)+n]
-        return cls(S, K, lmots_type)
+        I = buffer[4:4+LenI]
+        q = buffer[4+LenI:4+LenI+LenQ]
+        K = buffer[4+LenI+LenQ:(4+LenI+LenQ)+n]
+        return cls(I, q, K, lmots_type)
 
     @classmethod
     def sizeof(cls, buffer):
@@ -461,7 +470,7 @@ class LmotsPublicKey:
         if lmots_type not in lmots_params():
            raise ValueError(err_unknown_typecode, toHex(lmots_type))
         alg, n, p, w, ls = lmots_params[lmots_type]
-        rv = 4 + LenS + n
+        rv = 4 + LenI + LenQ + n
         if len(buffer) != rv:
             raise ValueError(err_bad_length, str(len(buffer)))
         return rv
@@ -469,7 +478,8 @@ class LmotsPublicKey:
     def prettyPrint(self):
         rv = "LMOTS public key\n"
         rv += ("   LMOTS type: %s\n" % toHex(self.type))
-        rv += ("   S         : %s\n" % toHex(self.S))
+        rv += ("   I         : %s\n" % toHex(self.I))
+        rv += ("   q         : %s\n" % toHex(self.q))
         rv += ("   K         : %s\n" % toHex(self.K))
         return rv
 
@@ -599,8 +609,7 @@ class LmsPrivateKey(object):
         self.ots_pub = []
         self._nodes = {}
         for j in range(0, 2**h):
-            S = self.I + u32(j)
-            priv = LmotsPrivateKey(S=S, SEED=self.SEED, lmots_type=lmots_type)
+            priv = LmotsPrivateKey(I=self.I, q=u32(j), SEED=self.SEED, lmots_type=lmots_type)
             pub = priv.publicKey()
             self.ots_priv.append(priv)
             self.ots_pub.append(pub)
@@ -708,8 +717,7 @@ class LmsPublicKey(object):
             return False
         if lms_sig.q > 2**h:
             return False
-        S = self.I + u32(lms_sig.q)
-        Kc = lms_sig.lmots_sig.buildPublic(S, message)
+        Kc = lms_sig.lmots_sig.buildPublic(self.I, u32(lms_sig.q), message)
         node_num = lms_sig.q + (2**h)
         tmp = H(alg, self.I + u32(node_num) + D_LEAF + Kc, m)
         for pv in lms_sig.path:

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -167,6 +167,7 @@ err_bad_value             = 'parameter has unknown value'
 err_bad_number_of_levels  = 'unsupported number of levels'
 err_private_key_exhausted = 'private key is exhausted'
 err_algorithm_mismatch    = 'LMOTS and LMS with different hash algorithms'
+err_prv_key_deserialize   = 'root private key not recognized'
 
 # ----------------------------------------------------------------------
 # The internal utility routines
@@ -869,7 +870,7 @@ class HssPrivateKey(object):
     """
     def __init__(self, levels=2, lms_type=lms_sha256_m32_h5,
                  lmots_type=lmots_sha256_n32_w8, SEED=None,
-                 remaining_signatures=None, prv0=None):
+                 remaining_signatures=None, prvs=None, sigs=None):
         if levels < 1 or levels > MaxHssLevels:
             ValueError(err_bad_number_of_levels, str(levels))
         if lmots_type not in lmots_params:
@@ -889,17 +890,22 @@ class HssPrivateKey(object):
             if len(SEED) != n:
                 raise ValueError(err_bad_length, str(len(SEED)))
             self.SEED = SEED
-        if prv0 is None:
-            prv0 = LmsPrivateKey(lms_type=lms_type,
-                                 lmots_type=lmots_type, SEED=SEED)
-        self.prv = [prv0]
+        if prvs is None:
+            prvs = [LmsPrivateKey(lms_type=lms_type,
+                                 lmots_type=lmots_type, SEED=SEED)]
+        self.prv = prvs
         if remaining_signatures is None:
             self._signatures_remaining = 2**(levels*h)
         else:
             self._signatures_remaining = remaining_signatures
-        self.pub = [prv0.publicKey()]
-        self.sig = []
-        for i in range(1, self.levels):
+        self.pub = []
+        for i in range(0, len(self.prv)):
+            self.pub.append(prvs[i].publicKey())
+        if sigs is None:
+            self.sig = []
+        else:
+            self.sig = sigs
+        for i in range(len(self.prv), self.levels):
             self.prv.append(LmsPrivateKey(
                 lms_type=lms_type, lmots_type=lmots_type, SEED=SEED))
             self.pub.append(self.prv[-1].publicKey())
@@ -943,8 +949,18 @@ class HssPrivateKey(object):
 
     def serialize(self):
         assert self._signatures_remaining.bit_length() <= 32, "_signatures_remaining=0x%x, %d bits"%(self._signatures_remaining,self._signatures_remaining.bit_length())
-        return u32(self.levels) + u32(self._signatures_remaining) + \
-               self.prv[0].serialize()
+        
+        rv = u32(self.levels) + u32(0)
+        serialized_prv0 = self.prv[0].serialize()
+        rv += u32(len(serialized_prv0)) + serialized_prv0
+
+        # all non-root keys and associated signatures
+        for i in range(1, len(self.prv)):
+            sig = self.sig[i - 1]
+            prv = self.prv[i]
+            serialized_prv = prv.serialize()
+            rv += u32(len(sig)) + sig + u32(len(serialized_prv)) + serialized_prv
+        return rv
 
     @classmethod
     def deserialize(cls, buffer):
@@ -952,9 +968,56 @@ class HssPrivateKey(object):
             raise ValueError(err_bad_length, str(len(buffer)))
         levels = int32(buffer[0:4])
         rs = int32(buffer[4:8])
+        if rs == 0:
+            return cls.deserializeV2(buffer)
         prv = LmsPrivateKey.deserialize(buffer[8:])
         return cls(levels, lms_type=prv.lms_type, lmots_type=prv.lmots_type, \
-                   remaining_signatures=rs, prv0=prv)
+                   remaining_signatures=rs, prvs=[prv])
+
+    @classmethod
+    def deserializeV2(cls, buffer):
+        """
+        levels - 4 bytes
+        padding of all zeros - 4 bytes
+        [root key_length in bytes + root key]
+        [signature_length in bytes + signature + key_length in bytes + key]... for each additional level's private key
+        """
+        if len(buffer) < 8:
+            raise ValueError(err_bad_length, str(len(buffer)))
+        levels = int32(buffer[0:4])
+        padding = int32(buffer[4:8])
+
+        remaining_signatures = 0
+        prvs = []
+        sigs = []
+        read_bytes = 8
+        while read_bytes < len(buffer):
+            if len(prvs) > 0:
+                # for every key after the root key
+                sig_length = int32(buffer[read_bytes:read_bytes+4])
+                read_bytes += 4
+                sig = buffer[read_bytes:read_bytes+sig_length]
+                read_bytes += sig_length
+                sigs.append(sig)
+            key_length = int32(buffer[read_bytes:read_bytes+4])
+            read_bytes += 4
+            prv = LmsPrivateKey.deserialize(buffer[read_bytes:read_bytes+key_length])
+            read_bytes += key_length
+            prvs.append(prv)
+
+            # Calculate how many remaining signatures this key can generate for us
+            # We need to account for how many levels of keys the current key can generate, and how much each use of a key is "worth"
+            # to be able to calculate how many effective signatures we've expended so far per key.
+            # Based off: 2**(self.levels*h)
+            alg2, m, h = lms_params[prv.lms_type]
+            remaining_signatures_from_key = 2**((levels-len(prvs)+1)*h) - ((2**((levels-len(prvs))*h)*(prv.maxSignatures()-prv.remaining())))
+            remaining_signatures += remaining_signatures_from_key
+        if len(prvs) < 1:
+            raise ValueError(err_prv_key_deserialize)
+        if len(sigs) < 1:
+            sigs = None
+        return cls(levels, lms_type=prvs[0].lms_type, lmots_type=prvs[0].lmots_type, \
+                   remaining_signatures=remaining_signatures, prvs=prvs, sigs=sigs)
 
     def prettyPrint(self):
         rv = "HSS private key\n"

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -387,7 +387,7 @@ class LmotsPrivateKey:
         hash = H_start(alg)
         H_update(hash, self.I + self.q + D_PBLC)
         for i in range(0, p):
-            tmp = H(alg, self.I + self.q + u16(i+1) + D_PRG + self.SEED, n)
+            tmp = H(alg, self.I + self.q + u16(i) + D_PRG + self.SEED, n)
             for j in range(0, (2**w)-1):
                 tmp = H(alg, self.I + self.q + u16(i) + u8(j) + tmp, n)
             H_update(hash, tmp)

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -402,7 +402,7 @@ class LmotsPrivateKey:
         V = hash1 + checksum(hash1, w, ls)
         y = []
         for i in range(0, p):
-            tmp = H(alg, self.I + self.q + u16(i+1) + D_PRG + self.SEED, n)
+            tmp = H(alg, self.I + self.q + u16(i) + D_PRG + self.SEED, n)
             for j in range(0, coef(V, i, w)):
                 tmp = H(alg, self.I + self.q + u16(i) + u8(j) + tmp, n)
             y.append(tmp)

--- a/pyhsslms/pyhsslms.py
+++ b/pyhsslms/pyhsslms.py
@@ -947,9 +947,7 @@ class HssPrivateKey(object):
         alg2, m, h = lms_params[self.lms_type]
         return 2**(self.levels*h)
 
-    def serialize(self):
-        # assert self._signatures_remaining.bit_length() <= 32, "_signatures_remaining=0x%x, %d bits"%(self._signatures_remaining,self._signatures_remaining.bit_length())
-        
+    def serialize(self):        
         # make sure to maintain 4-byte padding of all 1's to indicate we should use deserializeV2
         rv = u32(self.levels) + fromHex('1'*8)
         serialized_prv0 = self.prv[0].serialize()
@@ -1031,6 +1029,7 @@ class HssPrivateKey(object):
             rv += ("   SEED      : %s\n" % toHex(prv.SEED))
             rv += ("   q         : %s\n" % toHex(u32(prv.q)))
             rv += ("   pub       : %s\n" % toHex(prv.pub))
+        rv += ("   sigs left : %d\n" % self.remaining())
         rv += ("   max signs : %d\n" % self.maxSignatures())
         return rv
 

--- a/tests/test_hsslms.py
+++ b/tests/test_hsslms.py
@@ -39,7 +39,7 @@ import os
 import tempfile
 import unittest
 from pyhsslms import *
-from pyhsslms.compat import fromHex, toHex, toBytes, charNum, u8
+from pyhsslms.compat import fromHex, toHex, toBytes, charNum, u8, u32
 
 
 def mangle(buffer, offset=30):
@@ -167,6 +167,30 @@ class TestLMOTS(unittest.TestCase):
         self.assertIn('SEED      : 1e305866bfc4d18c4735bfc677711109', prvpp)
         pubpp = pub.prettyPrint()
         self.assertIn('I         : c6d47a98577cd2f13007908fd14309ca', pubpp)
+
+    def testPublicKeyGeneration(self):
+        I = fromHex('1'*32)
+        q = u32(123456789)
+        SEED = fromHex('3'*64)
+        lmots_type = lmots_sha256_n32_w2
+
+        S = I + q
+        prv = pyhsslms.LmotsPrivateKey(S, SEED, lmots_type)
+        pub = prv.publicKey()
+
+        expected_pub_bytes = fromHex('00000002' + \
+            # I - 16 bytes
+            '11111111111111111111111111111111' + \
+            # Q - 32-bit number
+            '075bcd15' + \
+            # KEY
+            '0cc6d68ead0190ad29b4125af6ca9ff1c4f23c4db4524411974653c9fb0c6e76')
+        expected_pub = pyhsslms.LmotsPublicKey.deserialize(expected_pub_bytes)
+
+        self.assertEqual(pub.type, expected_pub.type)
+        self.assertEqual(pub.S, expected_pub.S)
+        self.assertEqual(pub.K, expected_pub.K)
+        self.assertEqual(pub.serialize(), expected_pub.serialize())
 
     def testRandomPrivateKey(self):
         msg = toBytes('The way to get started is to quit talking and ' + \

--- a/tests/test_hsslms.py
+++ b/tests/test_hsslms.py
@@ -302,6 +302,28 @@ class TestHSS(unittest.TestCase):
         self.assertIn('I         : ee9a2418209ce10bc035de0f55c5eecf', pubpp)
         self.assertIn('K         : a3894ff49d2fa0e4a85b214f7c155901', pubpp)
 
+    def testExhaustedSign(self):
+        msg = toBytes('The way to get started is to quit talking and ' + \
+                      'begin doing.')
+        prv = pyhsslms.HssPrivateKey(levels=2)
+        self.assertEqual(prv.remaining(), 1024)
+        sigbuffer = prv.sign(msg)
+        self.assertEqual(prv.remaining(), 1023)
+    
+    def testSerializeDeserialize(self):
+        msg = toBytes('The way to get started is to quit talking and ' + \
+                      'begin doing.')
+        prv = pyhsslms.HssPrivateKey(levels=2)
+        self.assertEqual(prv.remaining(), 1024)
+        for i in range(0, 10):
+            prv.sign(msg)
+        self.assertEqual(prv.remaining(), 1014)
+        prv_serialized = prv.serialize()
+        prv_deserialized = pyhsslms.HssPrivateKey.deserialize(prv_serialized)
+        self.assertEqual(prv_deserialized.remaining(), 1014)
+        self.assertEqual(prv.prettyPrint(), prv_deserialized.prettyPrint())
+
+
     def testSmallRandomPrivateKey(self):
         msg = toBytes('The way to get started is to quit talking and ' + \
                       'begin doing.')

--- a/tests/test_hsslms.py
+++ b/tests/test_hsslms.py
@@ -174,8 +174,7 @@ class TestLMOTS(unittest.TestCase):
         SEED = fromHex('3'*64)
         lmots_type = lmots_sha256_n32_w2
 
-        S = I + q
-        prv = pyhsslms.LmotsPrivateKey(S, SEED, lmots_type)
+        prv = pyhsslms.LmotsPrivateKey(I, q, SEED, lmots_type)
         pub = prv.publicKey()
 
         expected_pub_bytes = fromHex('00000002' + \
@@ -188,7 +187,8 @@ class TestLMOTS(unittest.TestCase):
         expected_pub = pyhsslms.LmotsPublicKey.deserialize(expected_pub_bytes)
 
         self.assertEqual(pub.type, expected_pub.type)
-        self.assertEqual(pub.S, expected_pub.S)
+        self.assertEqual(pub.I, expected_pub.I)
+        self.assertEqual(pub.q, expected_pub.q)
         self.assertEqual(pub.K, expected_pub.K)
         self.assertEqual(pub.serialize(), expected_pub.serialize())
 

--- a/tests/test_hsslms.py
+++ b/tests/test_hsslms.py
@@ -93,6 +93,8 @@ class TestLMOTS(unittest.TestCase):
 
     def testBuildPublic(self):
         S = fromHex('61a5d57d37f5e46bfb7520806b07a1b800000005')
+        I = S[:pyhsslms.LenI]
+        q = S[pyhsslms.LenI:pyhsslms.LenI+pyhsslms.LenQ]
         msg = fromHex('0000000500000004d2f14ff6346af964569f7d6cb880a1b66' + \
          'c5004917da6eafe4d9ef6c6407b3db0e5485b122d9ebe15cda93cfec582d7ab')
         buf = fromHex( '00000004d32b56671d7eb98833c49b433c272586bc' + \
@@ -135,16 +137,18 @@ class TestLMOTS(unittest.TestCase):
         expected_pub = fromHex('87be83923a22106731e8f10f826faf4d02' + \
          '17b07d99694d1174d350fba7d578a1')
         sig = pyhsslms.LmotsSignature.deserialize(buf)
-        pub = sig.buildPublic(S, msg)
+        pub = sig.buildPublic(I, q, msg)
         self.assertEqual(expected_pub, pub)
 
     def testKnownPrivateKey(self):
         msg = toBytes('The way to get started is to quit talking and ' + \
                       'begin doing.')
         S = fromHex('c6d47a98577cd2f13007908fd14309ca00000001')
+        I = S[:pyhsslms.LenI]
+        q = S[pyhsslms.LenI:pyhsslms.LenI+pyhsslms.LenQ]
         seed = fromHex('1e305866bfc4d18c4735bfc677711109' + \
                              'a886656dc432e39281bc5a129b518172')
-        prv = pyhsslms.LmotsPrivateKey(S=S, SEED=seed)
+        prv = pyhsslms.LmotsPrivateKey(I=I, q=q, SEED=seed)
         self.assertEqual(1, prv.remaining())
         self.assertFalse(prv.is_exhausted())
         sigbuffer = prv.sign(msg)
@@ -159,10 +163,10 @@ class TestLMOTS(unittest.TestCase):
         self.assertTrue(sig.prettyPrint())
         prvpp = prv.prettyPrint()
         self.assertIn('LMOTS type: 00000004', prvpp)
-        self.assertIn('S         : c6d47a98577cd2f13007908fd14309ca', prvpp)
+        self.assertIn('I         : c6d47a98577cd2f13007908fd14309ca', prvpp)
         self.assertIn('SEED      : 1e305866bfc4d18c4735bfc677711109', prvpp)
         pubpp = pub.prettyPrint()
-        self.assertIn('S         : c6d47a98577cd2f13007908fd14309ca', pubpp)
+        self.assertIn('I         : c6d47a98577cd2f13007908fd14309ca', pubpp)
 
     def testRandomPrivateKey(self):
         msg = toBytes('The way to get started is to quit talking and ' + \
@@ -234,7 +238,8 @@ class TestLMS(unittest.TestCase):
         pub = prv.publicKey()
         pubserial = pub.serialize()
         pub2 = pyhsslms.LmotsPublicKey.deserialize(pubserial)
-        self.assertTrue(pub.S == pub2.S)
+        self.assertTrue(pub.I == pub2.I)
+        self.assertTrue(pub.q == pub2.q)
         self.assertTrue(pub.K == pub2.K)
         self.assertTrue(pub.type == pub2.type)
         self.assertTrue(pub.prettyPrint() == pub2.prettyPrint())

--- a/tests/test_hsslms.py
+++ b/tests/test_hsslms.py
@@ -236,10 +236,10 @@ class TestLMS(unittest.TestCase):
         self.assertIn('LMOTS type: 00000004', prvpp)
         self.assertIn('I         : e4a9fccb86dc9c49d71b72d5696acb54', prvpp)
         self.assertIn('SEED      : 0c0fe552dcf77d4bbe16b28605759ea4', prvpp)
-        self.assertIn('pub       : a8b35f4c521183a05d81fbcf8a81ffc9', prvpp)
+        self.assertIn('pub       : 7468c0058fc7e1dd87aef69fa5bbf1fc', prvpp)
         pubpp = pub.prettyPrint()
         self.assertIn('I         : e4a9fccb86dc9c49d71b72d5696acb54', pubpp)
-        self.assertIn('K         : a8b35f4c521183a05d81fbcf8a81ffc9', pubpp)
+        self.assertIn('K         : 7468c0058fc7e1dd87aef69fa5bbf1fc', pubpp)
 
     def testRandomPrivateKey(self):
         msg = toBytes('The way to get started is to quit talking and ' + \
@@ -300,7 +300,7 @@ class TestHSS(unittest.TestCase):
         pubpp = pub.prettyPrint()
         self.assertIn('levels    : 2', pubpp)
         self.assertIn('I         : ee9a2418209ce10bc035de0f55c5eecf', pubpp)
-        self.assertIn('K         : c584f571bbfdf285b9a8ac5bafb7738a', pubpp)
+        self.assertIn('K         : a3894ff49d2fa0e4a85b214f7c155901', pubpp)
 
     def testSmallRandomPrivateKey(self):
         msg = toBytes('The way to get started is to quit talking and ' + \

--- a/tests/test_hsslms.py
+++ b/tests/test_hsslms.py
@@ -302,6 +302,80 @@ class TestHSS(unittest.TestCase):
         self.assertIn('I         : ee9a2418209ce10bc035de0f55c5eecf', pubpp)
         self.assertIn('K         : a3894ff49d2fa0e4a85b214f7c155901', pubpp)
 
+    def testKnownV2PrivateKey(self):
+        msg = toBytes('The way to get started is to quit talking and ' + \
+                      'begin doing.')
+        buffer = fromHex('00000002111111110000003c0000000500000004f43d39dc2a0dbf22ede8' + \
+            '66b071682c30bdca160d3856d74b9c08be654397dd5db1f91ab94b7d0ec0' + \
+            '4fa184b698ede50c000000010000050c0000000000000004a80d9f386d96' + \
+            'd5c0e70c42e8e19143275f61f83c66fc504a1e57481fae3b3e44ad7b952f' + \
+            '3f4c491a2f47e78d92755bde9c25adb7cbf91e5f3e201d452e6ddb47360f' + \
+            '0bf03b06dd9f1110df25fa291a17f2a9cdd7ca1877b139764342f568bd87' + \
+            '44b7a1cf8e3c7c705c38958b4b41ff803206627f45c614648af55dca3ef5' + \
+            '83daf45eb92f63675709d0d9ec861d0a265ce25f3858dd23cc537fdf85fe' + \
+            'd0cb233aa5df3b8439c7bd031be632dda2d06076c150f2fed1d7da7ce2e3' + \
+            '1c856fe5676a8aa6a3459e75a5ea6154694d5eb9299defda07139d46d279' + \
+            'a1d38d0381f61b899dca2666a78b992945ba6d2fba5d64364d562258c3fb' + \
+            '7962ddc641f7856bdade24f3db10495deb7e648ca20a19ad5f10432aaaae' + \
+            '2dbd42358c8505fca3b948f461e890129991959374e12f125eb4328086fa' + \
+            'a72025e01a5ca9b72e59cb6764193dc5bd84ab09584a4200391fb218244e' + \
+            '1db3e3570821b52489e164287e260b05bb5c08146ec790d3b0fe8a06a5d6' + \
+            'd7c02a8d9d7b1d78d46cee47f7fea98c59ac270335a0b584f4a4d66a799a' + \
+            'a8e7afe56fcdd22fcf5697b7bf846b6b2c6fe928d69d75b6bdc18419410d' + \
+            '93cd6eda972dfd9ef71d64c31117c85d712f5eda7a6db75c4da5a80bbdbb' + \
+            '43b7db062a79357685e8177a6edf01b60a709b75f3745f1759e10807aeb6' + \
+            '06d233ba458958b18e9d08435d81efcb784e201ca16799911d272d8b692b' + \
+            '0dd6ad2fa22f78bc6dc3a3c92b839c942a84b75187f7a3ed6ad4068aaee8' + \
+            '80a43daa25010e75daa5afeeecc709a78ab0852b7d1c2d037419510591a3' + \
+            'd726a238715711a00b84c9f9d45ffc39654096baf6e8e24f3c3c1e419f25' + \
+            '23394d0b93d190e0636cdcbb3e5d0d7d698877d3e01b28d5b82919b0d1ca' + \
+            'e14991b7ac7f743bd6328a415840f7582debc2657b31fdb930980b8420fd' + \
+            '65e3e7ff84459e17663c357f4554ffc678de9c3bc6ec5f482fb21af7cbd9' + \
+            'd039fa82c6fef8cb07b97c31dba402541391a2619753e506dedbf84434b3' + \
+            '323d3e1cfde4ec86301b7d8dbdb585dfd2da0e4c613ff687af52fa6e568f' + \
+            '61be757bce7745c93bbeb35539a984056f45d2c4961828e15cfa0716e37f' + \
+            '697ed59eec11e72e77fdee7e1a0256d88d3b5128478627aefab10491c1c7' + \
+            'f1bee592f93d3d91be68d257a56a5166304bd0afaed0418c0faa5eb1d044' + \
+            '96d31ae6aac9d9dd9efa119b944255e7ab4738d93a36d1a6ae0fbb0b70ef' + \
+            '1bd176215a1799a8634f48f854cfb5248dab2b5e82798d7187571656f9a8' + \
+            'a7e00bceeb736e42b05abee3976a2346fda2848384dc13e5150b8f7adb55' + \
+            'eff76013194459704626e5ea5bf1316c763ad5180977a3ccd92b53f2e162' + \
+            '317465fb2efb60e463c22d76e63db4d6d88130efc69dfbd0957c37d56dfe' + \
+            'c90045b18ca12f45da06c085a279b667793fed11977d79893b38260fd6e3' + \
+            'aec03532f9db500634115eb0a123d9c849eee14298cb940061c2ae8d86c1' + \
+            'a10d0109ceffd767552df22d04507f57edfbeb983afb6c72e2614c7f3621' + \
+            'b16112fef7cd282cdcbd9092cad4bdf347da06a16f5c939effd69467eb85' + \
+            '34823e13000000054ca89ad451211b67dc722c9f954f6ea9be3957853267' + \
+            '19ce7c4f51bfde733e3375c0bcdf3e95d98246bcc47b30af65b6625b83f8' + \
+            'c7f4d9c4501f9fbcac45ceb6bdd8aaaa79d7e1dd4cf79f9780c73378372d' + \
+            'bb6d508bdf444457dc1697cd164dfe9a48b19fa3feff4350dbbf2d902aa3' + \
+            '14b0a0a6dfbd2ea53184664f6741656baf751a68c9fa3954a3508a8c01ba' + \
+            'da7358ce41bf47a7764902dccf137928b6e00000003c0000000500000004' + \
+            '1adbe10213248e5e2ef905276ec18a35fb3aac2a0e7692ad6ccf517bd984' + \
+            'be61e6540e52a69bc44de051c2d20c056a1600000000')                        
+        prv = pyhsslms.HssPrivateKey.deserialize(buffer)
+        self.assertEqual(1024, prv.remaining())
+        sigbuffer = prv.sign(msg)
+        self.assertEqual(2644, len(sigbuffer))
+        self.assertEqual(1023, prv.remaining())
+        self.assertFalse(prv.is_exhausted())
+        sig = pyhsslms.HssSignature.deserialize(sigbuffer)
+        pub = prv.publicKey()
+        self.assertTrue(pub.verify(msg, sigbuffer))
+        self.assertFalse(pub.verify(msg, mangle(sigbuffer)))
+        self.assertFalse(pub.verify(mangle(msg), sigbuffer))
+        self.assertTrue(sig.prettyPrint())
+        prvpp = prv.prettyPrint()
+        self.assertIn('levels    : 2', prvpp)
+        self.assertIn('LMS type  : 00000005', prvpp)
+        self.assertIn('LMOTS type: 00000004', prvpp)
+        self.assertIn('q         : 00000001', prvpp)
+        self.assertIn('I         : b1f91ab94b7d0ec04fa184b698ede50c', prvpp)
+        pubpp = pub.prettyPrint()
+        self.assertIn('levels    : 2', pubpp)
+        self.assertIn('I         : b1f91ab94b7d0ec04fa184b698ede50c', pubpp)
+        self.assertIn('K         : da0d59a80f460f17693c6bec1c6711e0', pubpp)
+
     def testExhaustedSign(self):
         msg = toBytes('The way to get started is to quit talking and ' + \
                       'begin doing.')
@@ -315,12 +389,12 @@ class TestHSS(unittest.TestCase):
                       'begin doing.')
         prv = pyhsslms.HssPrivateKey(levels=2)
         self.assertEqual(prv.remaining(), 1024)
-        for i in range(0, 10):
+        for i in range(0, 64):
             prv.sign(msg)
-        self.assertEqual(prv.remaining(), 1014)
+        remaining_sigs = prv.remaining()
         prv_serialized = prv.serialize()
         prv_deserialized = pyhsslms.HssPrivateKey.deserialize(prv_serialized)
-        self.assertEqual(prv_deserialized.remaining(), 1014)
+        self.assertEqual(prv_deserialized.remaining(), remaining_sigs)
         self.assertEqual(prv.prettyPrint(), prv_deserialized.prettyPrint())
 
 

